File: cpp/leetcode/RemoveComments.cpp
//Leetcode Problem No 722 Remove Comments
//Solution written by Xuqiang Fang on 30 May, 2018

#include <iostream>
#include <vector>
#include <string>
#include <algorithm>
#include <unordered_map>
#include <unordered_set>
#include <stack>
#include <queue>
#include <regex>

using namespace std;
class Solution{
public:
    vector<string> removeComments(vector<string>& source) {
        /*
        * only a few cases, double slash, block
        */
        vector<string> ans;        
        bool in = false;
        string b;
        for(auto&s : source){
            for(int j=0; j<s.length(); ++j){
                if(in){
                    if(s[j] == '*' && j < s.length() - 1 && s[j+1] == '/'){
                        in = false;
                        j++;
                    }
                } 
                else{
                    if(s[j] == '/' && j < s.length() - 1 && s[j+1] == '*'){
                        in = true;
                        j++;
                    }
                    else if(s[j] == '/' && j < s.length() - 1 && s[j+1] == '/'){
                        break;
                    }
                    else{
                        b.push_back(s[j]);
                    }
                }
            }
            if(!in && b.length() > 0){
                ans.push_back(b);
                b = "";
            }
        }
        return ans;
    }
};

int main(){
    Solution s;
    vector<string> s1{"/*Test program */", "int main()", "{ ", "  // variable declaration ", "aaa//a"};
    vector<string> s2{"a/*comment", "line", "more_comment*/b"};
    vector<string> s3{"/*Test program */", "int main()", "{ ", "  // variable declaration ", "int a, b, c;", "/* This is a test", "   multiline  ", "   comment for ", "   testing */", "a = b + c;", "}"};
    vector<string> ans = s.removeComments(s1);
    for(auto& a : ans){
        cout << a << endl;
    }
    ans = s.removeComments(s2);
    for(auto& a : ans){
        cout << a << endl;
    }
    ans = s.removeComments(s3);
    for(auto& a : ans){
        cout << a << endl;
    }
    return 0;
}

File: java/leetcode/RemoveComments.java
//Leetcode problem 722 Remove Comments
//Solution written by Xuqiang Fang on 31 May, 2018
import java.util.ArrayList;
import java.util.List;
import java.util.HashMap;
import java.util.Map;
import java.util.HashSet;
import java.util.Set;
import java.util.PriorityQueue;
import java.util.Stack;
import java.util.Queue;
import java.util.LinkedList;
import java.util.Arrays;
class Solution{
    public List<String> removeComments(String[] source) {
        List<String> ans = new ArrayList<>();
        boolean in = false;
        StringBuilder sb = new StringBuilder();
        for(String s : source){
            for(int i=0; i<s.length(); ++i){
                if(in){
                    if(s.charAt(i) == '*' && i < s.length()-1 && s.charAt(i+1) == '/'){
                        in = false;
                        i++;
                    }
                }
                else{
                    if(s.charAt(i) == '/' && i < s.length()-1 && s.charAt(i+1) == '/'){
                        break;
                    }
                    else if(s.charAt(i) == '/' && i < s.length()-1 && s.charAt(i+1) == '*'){
                        in = true;
                        i++;
                    }
                    else{
                        sb.append(s.charAt(i));
                    }
                }
            } 
            if(!in && sb.length() > 0){
                ans.add(sb.toString());
                sb.delete(0, sb.length());
            }
        }
        return ans;
    }
}

public class RemoveComments{
	public static void main(String[] args){
		Solution s = new Solution();
        String[] source = {"/*Test program */", "int main()", "{ ", "  // variable declaration ", "int a, b, c;", "/* This is a test", "   multiline  ", "   comment for ", "   testing */", "a = b + c;", "}"};
        System.out.println(s.removeComments(source));
	}
}
